src/dataflow/src/render/join/delta_join.rs

```diff
@@ -562,43 +516,35 @@ where
             let row_key = row_packer.finish_and_reuse();
             // Explicit drop to release borrow on `row` so that it can be returned.
             drop(datums_local);
-            Ok((row, row_key))
+            Ok((row_key, row, time))
         }
     });
 
     use differential_dataflow::AsCollection;
     use timely::dataflow::operators::OkErr;
 
     let mut datums = DatumVec::new();
-    let (oks, errs2) = dogsdogsdogs::operators::lookup_map(
+    let (oks, errs2) = dogsdogsdogs::operators::half_join(
```

Comment:
I just got the following panic with a randomized DBT/TCH workload .  I am trying to reproduce but it does not seem reproducible. Any pointers as to how this could happen so that I can try for a more tailored test case?


```
Operator prematurely shut down: HalfJoin
  true
  [ChangeBatch { updates: [(1617781501344, 1)], clean: 1 }, ChangeBatch { updates: [], clean: 0 }]
Operator prematurely shut down: HalfJoin
  true
  [ChangeBatch { updates: [(1617781501344, 1)], clean: 1 }, ChangeBatch { updates: [], clean: 0 }]
Operator prematurely shut down: HalfJoin
  true
  [ChangeBatch { updates: [(1617781501344, 1)], clean: 1 }, ChangeBatch { updates: [], clean: 0 }]
Operator prematurely shut down: HalfJoin
  true
  [ChangeBatch { updates: [(1617781501344, 1)], clean: 1 }, ChangeBatch { updates: [], clean: 0 }]
Operator prematurely shut down: HalfJoin
  true
  [ChangeBatch { updates: [(1617781501344, 1)], clean: 1 }, ChangeBatch { updates: [], clean: 0 }]
Operator prematurely shut down: HalfJoin
  true
  [ChangeBatch { updates: [(1617781501344, 1)], clean: 1 }, ChangeBatch { updates: [], clean: 0 }]
Operator prematurely shut down: HalfJoin
  true
  [ChangeBatch { updates: [(1617781501344, 1)], clean: 1 }, ChangeBatch { updates: [], clean: 0 }]
2021-04-07T07:45:01.446041Z ERROR panic: explicit panic
thread: timely:work-7
location: /home/pstoev/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/0d0d848/timely/src/progress/subgraph.rs:678:17
version: 0.7.2-dev (b04a9d6d1b8a7c90c6d0ec7202774c72502c7f7c)
backtrace:
   0: materialized::handle_panic
             at src/materialized/src/bin/materialized/main.rs:652:21
   1: core::ops::function::Fn::call
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/core/src/ops/function.rs:70:5
   2: std::panicking::rust_panic_with_hook
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:595:17
   3: std::panicking::begin_panic::{{closure}}
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:520:9
   4: std::sys_common::backtrace::__rust_end_short_backtrace
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/sys_common/backtrace.rs:141:18
   5: std::panicking::begin_panic
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:519:12
   6: timely::progress::subgraph::PerOperatorState<T>::schedule
             at /home/pstoev/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/0d0d848/timely/src/progress/subgraph.rs:678:17
   7: timely::progress::subgraph::Subgraph<TOuter,TInner>::activate_child
             at /home/pstoev/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/0d0d848/timely/src/progress/subgraph.rs:344:26
      <timely::progress::subgraph::Subgraph<TOuter,TInner> as timely::scheduling::Schedule>::schedule
             at /home/pstoev/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/0d0d848/timely/src/progress/subgraph.rs:310:17
   8: timely::progress::subgraph::PerOperatorState<T>::schedule
             at /home/pstoev/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/0d0d848/timely/src/progress/subgraph.rs:662:30
   9: timely::progress::subgraph::Subgraph<TOuter,TInner>::activate_child
             at /home/pstoev/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/0d0d848/timely/src/progress/subgraph.rs:344:26
      <timely::progress::subgraph::Subgraph<TOuter,TInner> as timely::scheduling::Schedule>::schedule
             at /home/pstoev/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/0d0d848/timely/src/progress/subgraph.rs:310:17
  10: timely::progress::subgraph::PerOperatorState<T>::schedule
             at /home/pstoev/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/0d0d848/timely/src/progress/subgraph.rs:662:30
  11: timely::progress::subgraph::Subgraph<TOuter,TInner>::activate_child
             at /home/pstoev/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/0d0d848/timely/src/progress/subgraph.rs:344:26
      <timely::progress::subgraph::Subgraph<TOuter,TInner> as timely::scheduling::Schedule>::schedule
             at /home/pstoev/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/0d0d848/timely/src/progress/subgraph.rs:310:17
  12: timely::progress::subgraph::PerOperatorState<T>::schedule
             at /home/pstoev/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/0d0d848/timely/src/progress/subgraph.rs:662:30
  13: timely::progress::subgraph::Subgraph<TOuter,TInner>::activate_child
             at /home/pstoev/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/0d0d848/timely/src/progress/subgraph.rs:344:26
      <timely::progress::subgraph::Subgraph<TOuter,TInner> as timely::scheduling::Schedule>::schedule
             at /home/pstoev/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/0d0d848/timely/src/progress/subgraph.rs:310:17
  14: timely::worker::Wrapper::step::{{closure}}
             at /home/pstoev/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/0d0d848/timely/src/worker.rs:747:57
      core::option::Option<T>::map
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/core/src/option.rs:487:29
      timely::worker::Wrapper::step
             at /home/pstoev/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/0d0d848/timely/src/worker.rs:747:26
  15: timely::worker::Worker<A>::step_or_park
             at /home/pstoev/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/0d0d848/timely/src/worker.rs:395:38
  16: dataflow::server::Worker<A>::run
             at src/dataflow/src/server.rs:424:13
  17: dataflow::server::serve::{{closure}}
             at src/dataflow/src/server.rs:206:13
      timely::execute::execute::{{closure}}
             at /home/pstoev/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/0d0d848/timely/src/execute.rs:285:22
      timely_communication::initialize::initialize_from::{{closure}}
             at /home/pstoev/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/0d0d848/communication/src/initialize.rs:285:33
      std::sys_common::backtrace::__rust_begin_short_backtrace
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/sys_common/backtrace.rs:125:18
  18: std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/thread/mod.rs:474:17
      <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panic.rs:344:9
      std::panicking::try::do_call
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:379:40
      std::panicking::try
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:343:19
      std::panic::catch_unwind
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panic.rs:431:14
      std::thread::Builder::spawn_unchecked::{{closure}}
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/thread/mod.rs:473:30
      core::ops::function::FnOnce::call_once{{vtable.shim}}
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/core/src/ops/function.rs:227:5
  19: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/alloc/src/boxed.rs:1521:9
      <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/alloc/src/boxed.rs:1521:9
      std::sys::unix::thread::Thread::new::thread_start
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/sys/unix/thread.rs:71:17
  20: start_thread
  21: clone
```

Reply:
This is an internal integrity check in timely dataflow, which is testing that once we close an operator (which only happens when we are certain no new records will arrive) we should never again have changes to the frontier of timestamps that might arrive.

What the report says with
```
Operator prematurely shut down: HalfJoin
  true
  [ChangeBatch { updates: [(1617781501344, 1)], clean: 1 }, ChangeBatch { updates: [], clean: 0 }]
```
is that `HalfJoin` (certainly suspicious as the new operator here) 1. asks to be told of changes to its incoming timestamps (the `true`) and that it has received a change to the frontier of its first input having already been shut down.

The change, `ChangeBatch { updates: [(1617781501344, 1)], clean: 1 }`, seems to be out of place, as it is a strict addition to what was there before, which isn't supposed to happen except on start-up (frontier changes are meant to be "downgrades", introducing new timestamps only if a strictly prior timestamp is deleted at the same time).

This makes me think there may be a race in start-up, that e.g. shuts down an operator before it can get started, and then errors when it attempts to supply the initial statement about the initial frontiers for the operator. This code was recently changed in response to a different bug you found, though ... that fix would have been the sort of thing that would have prevented this error, I believe.

In any case, I'll start to take a peek, but if you find it reproducing more often in some setting I'd be interested.

There are seven errors reported up above; how many workers were you using at the time? (i.e. is it "all of them")

Reply:
I should say more explicitly: this error is not just a "this PR has a bug" but "this PR shows off a bug in timely dataflow" which is much more significant.

Reply:
I had 16 workers I belive. Certainly not 7.

Reply:
I got this with a different operator mentioned:

```
Operator prematurely shut down: delta path 1
  true
  [ChangeBatch { updates: [(1617863968370, 1)], clean: 1 }, ChangeBatch { updates: [], clean: 0 }, ChangeBatch { updates: [], clean: 0 }]
Operator prematurely shut down: delta path 1
  true
  [ChangeBatch { updates: [(1617863968370, 1)], clean: 1 }, ChangeBatch { updates: [], clean: 0 }, ChangeBatch { updates: [], clean: 0 }]
Operator prematurely shut down: delta path 1
  true
  [ChangeBatch { updates: [(1617863968370, 1)], clean: 1 }, ChangeBatch { updates: [], clean: 0 }, ChangeBatch { updates: [], clean: 0 }]
Operator prematurely shut down: delta path 1
  true
  [ChangeBatch { updates: [(1617863968370, 1)], clean: 1 }, ChangeBatch { updates: [], clean: 0 }, ChangeBatch { updates: [], clean: 0 }]
Operator prematurely shut down: delta path 1
  true
  [ChangeBatch { updates: [(1617863968370, 1)], clean: 1 }, ChangeBatch { updates: [], clean: 0 }, ChangeBatch { updates: [], clean: 0 }]
```
@frankmcsherry does that mean that I should continue fishing for this panic in "main" since the half join thing is not involved?

Reply:
Here is a test case that should cause the crash within a minute. Run with --workers 4, and pipe it through the psql client, sending all the output to /dev/null as to not delay the test.

[Operator-prematurely-shut-down.sql.gz](https://github.com/MaterializeInc/materialize/files/6278433/Operator-prematurely-shut-down.sql.gz)

It feels to me this is some race condition that requires the following to happen:
- workers  > 1 but less than 32
- inserts arriving in the base tables used by the view

Reply:
Thanks very much! I may pop in with questions, but I appreciate the repro and will get started on it.

Reply:
I am back with questions! I'm running it like so
```
mcsherry@gallustrate materialize % time cat ~/Downloads/Operator-prematurely-shut-down.sql |  psql -h localhost -p 6875 -d materialize -U materialize > /dev/null
ERROR:  unknown catalog item 'v1914018'
ERROR:  unknown catalog item 'v1914018'
ERROR:  unknown catalog item 'v1914018'
ERROR:  unknown catalog item 'v1914018'
cat ~/Downloads/Operator-prematurely-shut-down.sql  0.00s user 0.01s system 0% cpu 21.698 total
psql -h localhost -p 6875 -d materialize -U materialize > /dev/null  0.22s user 0.25s system 1% cpu 26.397 total
mcsherry@gallustrate materialize % 
```
It seems to complete in about 20 seconds, and hasn't yet (four runs) exhibited a fault. Is your experience different? Should I be running it multiple times and it is just luck whether it happens for you?

Reply:
Ah, I've got it going. Nevermind, and thank you!

Reply:
@frankmcsherry do you require separate test cases for ```Operator prematurely shut down: delta path 1``` and for ```Operator prematurely shut down: HalfJoin``` or it is the same underlying issue?

Reply:
I suspect it is the same underlying issue. I think I have what I need at the moment (it does repro now, not deterministically but enough to be actionable!).